docs/proposals/20221205-cpu-schedule-latency.md

```diff
@@ -0,0 +1,240 @@
+
+#### eBPF
+
+The eBPF program use belowing three BPF Maps to complete the working process. 
+
+```c
+struct {
+	__uint(type, BPF_MAP_TYPE_CGROUP_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, u32);
+} cgroup SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32);
+	__type(value, struct rqeventv2);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} output SEC(".maps");
+```
+
+The `BPF_MAP_TYPE_CGROUP_ARRAY` map `cgroup` is used to store `cgroup file descriptors` in user-space for later use to call some helper functions. To collect CPU schedule latency on per `Pod` or `Container` dimension, we use a helper function `bpf_current_task_under_cgroup()` provided by Linux kernel `bpf_helpers` as a filter. It has the ability to check whether the probe is being run is the context of a given subset of the cgroup.  
+
+When events of probe `tp/sched/sched_wakeup` or `tp/sched/sched_wakeup_new` arrive, the current timestamp need to be recorded and stored in the `BPF_MAP_TYPE_HASH` map `start` with key of its `pid` . And when events of probe `tp/sched/sched_switch` arrive, we try to get elements from map `start` by the same key `pid`. If it exists, the current timestamp is subtracted by the start timestamp as one record of schedule latency. This delta is put into the `BPF_MAP_TYPE_PERF_EVENT_ARRAY` map which is further read by user-space go program. 
+
+To implement the above process, some other structure need to be defined to catch and handle corresponding events. 
+
+Belowing `sched_wakeup_args` and `sched_switch_args` is defined as the format in `/sys/kernel/debug/tracing/events/sched/{$probe_name}/format` file. And `trace_entry` is the common part of them. 
+
+```c
+struct trace_entry {
+	short unsigned int type;
+	unsigned char flags;
+	unsigned char preempt_count;
+	int pid;
+};
+
+struct sched_wakeup_args {
+	struct trace_entry ent;
+	char comm[16];
+	pid_t pid;
+	int prio;
+	int success;
+	int target_cpu;
+	char __data[0];
+};
+
+struct sched_switch_args {
+	struct trace_entry ent;
+	char prev_comm[16];
+	pid_t prev_pid;
+	int prev_prio;
+	long int prev_state;
+	char next_comm[16];
+	pid_t next_pid;
+	int next_prio;
+	char __data[0];
+};
+```
+
+Then we need to implement belowing three handle functions to trace these tracepoints. 
+
+```c
+SEC("tp/sched/sched_wakeup")
+int handle__sched_wakeup(struct sched_wakeup_args *ctx)
+{
+	return 0;
+}
+
+SEC("tp/sched/sched_wakeup_new")
+int handle__sched_wakeup_new(struct sched_wakeup_args *ctx)
+{
+  return 0;
+}
+
+SEC("tp/sched/sched_switch")
+int handle_switch(struct sched_switch_args *ctx)
+{
+	return 0;
+}
+```
+
+#### Use in Go
```

Comment:
like we discussed before, how to implant the process into different environments.

Reply:
The proposal is now adjusted to avoid this problem